src/fast-terms.jl

```diff
@@ -0,0 +1,278 @@
+import Base: +, -, *, /, ^
+
+const SN = Symbolic{<:Number}
+"""
+    Add(coeff, dict)
+
+Represents coeff + (key1 * val1) + (key2 * val2) + ...
+
+where coeff and the vals are non-symbolic numbers.
+
+"""
+struct Add{X, T, D} <: Symbolic{X}
+    coeff::T
+    dict::D
+end
+
+function Add(coeff, dict)
+    if isempty(dict)
+        return coeff
+    end
+    Add{Number, typeof(coeff), typeof(dict)}(coeff,dict)
+end
+
+symtype(a::Add{X}) where {X} = X
+
+istree(a::Add) = true
+
+operation(a::Add) = +
+
+arguments(a::Add) = vcat(a.coeff, [v*k for (k,v) in a.dict])
+
+Base.hash(a::Add, u::UInt64) = hash(a.coeff, hash(a.dict, u))
+
+Base.isequal(a::Add, b::Add) = isequal(a.coeff, b.coeff) && isequal(a.dict, b.dict)
+
+function Base.show(io::IO, a::Add)
+    print_coeff = !iszero(a.coeff)
+    print_coeff && print(io, a.coeff)
+
+    for (i, (k, v)) in enumerate(a.dict)
+        if (i == 1 && print_coeff) || i != 1
+            print(io, " + ")
+        end
+        if isone(v)
+            print(io, k)
+        else
+            print(io, v, k)
+        end
+    end
+end
+
+"""
+make_add_dict(sign, xs...)
+
+Any Muls inside an Add should always have a coeff of 1
+and the key (in Add) should instead be used to store the actual coefficient
+"""
+function make_add_dict(sign, xs...)
+    d = Dict{Any, Number}()
+    for x in xs
+        if x isa Mul
+            k = Mul(1, x.dict)
+            v = sign * x.coeff + get(d, k, 0)
+        else
+            k = x
+            v = sign + get(d, x, 0)
+        end
+        if iszero(v)
+            delete!(d, k)
+        else
+            d[k] = v
+        end
+    end
+    d
+end
+
++(a::Number, b::SN) = Add(a, make_add_dict(1, b))
+
++(a::SN, b::Number) = Add(b, make_add_dict(1, a))
+
+function +(a::SN, b::SN)
+    if a isa Add
+        return a + Add(0, make_add_dict(1, b))
+    elseif b isa Add
+        return b + a
+    end
+    Add(0, make_add_dict(1, a, b))
+end
+
++(a::Add, b::Add) = Add(a.coeff + b.coeff, _merge(+, a.dict, b.dict, filter=iszero))
+
++(a::Number, b::Add) = iszero(a) ? b : Add(a, make_add_dict(1, b))
+
++(b::Add, a::Number) = iszero(a) ? b : Add(a, make_add_dict(1, b))
```

Comment:
This may not always be an Add. For eg: `(x-1) + 1` should not be an Add. If you keep it as an Add, then expressions will get more and more complicated.

Reply:
Should it be
```suggestion
+(b::Add, a::Number) = iszero(a) ? b : Add(a + b.coeff, b.dict)
```
?

Reply:
That's one optimization. Another is if `a + b.coeff == 0 and len(b.dict) == 1` then it should not be an Add.